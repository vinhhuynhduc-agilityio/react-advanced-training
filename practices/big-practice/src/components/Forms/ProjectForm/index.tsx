import { useForm } from 'react-hook-form';

// components
import { Button } from '@/components/common';
import { TextField } from '@/components';

// helpers
import { isProjectDuplicate } from './helpers';

// types
import { ProjectsData } from '@/types';

interface ProjectFormProps {
  onClose: () => void;
  onSubmit: (newProjectName: string) => void;
  projects: ProjectsData[]
};

interface ProjectFormData {
  projectName: string;
};

const ProjectForm: React.FC<ProjectFormProps> = ({
  onClose,
  onSubmit,
  projects
}) => {

  const {
    register,
    handleSubmit,
    formState: { errors }
  } = useForm<ProjectFormData>();

  const onSubmitForm = (data: ProjectFormData) => {
    onSubmit(data.projectName);
  };

  return (
    <form
      onSubmit={handleSubmit(onSubmitForm)}
      className="space-y-4"
      data-testid="project-form"
    >

      {/* Body */}
      <TextField
        name='projectName'
        label='Project Name'
        placeholder='Enter project name'
        labelWidth='w-32'
        register={register}
        withErrorMargin='ml-32'
        validation={{
          required: 'Project name is required',
          validate: {
            duplicate: (value) =>
              !isProjectDuplicate(projects, value) || 'Project name already exists',
          },
        }}
        error={errors.projectName?.message}
      />

      {/* Footer */}
      <div className="flex justify-between border-t pt-3">
        <Button
          type='button'
          onClick={onClose}
          label='Cancel'
          variant='secondary'
        />
        <Button
          type='submit'
          label='Save'
          variant='primary'
        />
      </div>
    </form>
  );
};

export default ProjectForm;
