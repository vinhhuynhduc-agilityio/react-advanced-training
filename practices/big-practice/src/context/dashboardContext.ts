import { createContext } from 'react';

import {
  ProjectsData,
  TaskData,
  UserData
} from '@/types';

interface DashboardContextType {
  users: UserData[];
  tasks: TaskData[];
  projects: ProjectsData[];
  setUsers: React.Dispatch<React.SetStateAction<UserData[]>>;
  setTasks: React.Dispatch<React.SetStateAction<TaskData[]>>;
  setProjects: React.Dispatch<React.SetStateAction<ProjectsData[]>>;
}

export const DashboardContext = createContext<DashboardContextType | null>(null);
