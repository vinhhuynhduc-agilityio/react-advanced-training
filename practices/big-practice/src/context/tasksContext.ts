import { createContext } from 'react';

import {
  TaskData,
} from '@/types';

interface TasksContextType {
  tasks: TaskData[];
  setTasks: React.Dispatch<React.SetStateAction<TaskData[]>>;
}

export const TasksContext = createContext<TasksContextType | null>(null);
