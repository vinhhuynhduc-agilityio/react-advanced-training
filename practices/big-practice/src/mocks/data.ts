export const mockUsers = [
  {
    id: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    fullName: "Joe Bloggs",
    earnings: "$11500",
    email: "john@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=1",
    registered: "May 21, 2020 17:02:06",
    lastUpdated: "Nov 11, 2024 14:34:21",
  },
  {
    id: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    fullName: "Jane Smith",
    earnings: "$36911",
    email: "jane@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=2",
    registered: "June 10, 2021 09:20:15",
    lastUpdated: "Nov 18, 2024 17:33:37",
  },
  {
    id: "123e4567-e89b-12d3-a456-426614174000",
    fullName: "Alice Brown",
    earnings: "$7200",
    email: "alice@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=3",
    registered: "April 15, 2020 14:02:06",
    lastUpdated: "October 11, 2023 10:00:00",
  },
];

export const mockProject = [
  {
    id: "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
    projectName: "Support",
  },
  {
    id: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    projectName: "Failure Testing",
  },
  {
    id: "e38e56a2-4d3c-469d-8345-45d6b5fae9f9",
    projectName: "Quality Management",
  },
  {
    id: "f49b68f4-5e5b-457d-bf16-1b5b78e6f5f1",
    projectName: "Data Quality",
  },
  {
    id: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    projectName: "Test",
  },
];

export const mockTasks = [
  {
    id: "71e564f4-7c18-47f7-89f2-abe4b7ec2854",
    userId: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    projectId: "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
    taskName: "Build test",
    startDate: "10 Aug 24",
    completedDate: "15 Nov 24",
    currency: 2000,
    status: true,
    projectName: "Support",
    fullName: "Joe Bloggs",
  },
  {
    id: "82e564f4-7c18-47f7-89f2-abe4b7ec2856",
    userId: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    projectId: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    taskName: "Build Frontend",
    startDate: "05 Sep 23",
    completedDate: "27 Oct 24",
    currency: 4000,
    status: true,
    projectName: "Failure Testing",
    fullName: "Joe Bloggs",
  },
  {
    id: "u2e564f4-7c18-47f7-89f2-abe4b7ec2878",
    userId: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    projectId: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    taskName: "Develop Chatbot Feature",
    startDate: "23 Sep 23",
    completedDate: "24 Oct 24",
    currency: 3200,
    status: true,
    projectName: "Failure Testing",
    fullName: "Jane Smith",
  },
  {
    id: "7ce698fc-c0ee-4d8c-8a1b-df988a96c22f",
    userId: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    projectId: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    taskName: "test today",
    startDate: "13 Nov 24",
    completedDate: "incomplete",
    currency: 1111,
    status: false,
    projectName: "Test",
    fullName: "Jane Smith",
  },
  {
    id: "r0s1t2u3-7c18-47f7-89f2-abe4b7ec2949",
    userId: "123e4567-e89b-12d3-a456-426614174000",
    projectId: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    taskName: "Data Migration",
    startDate: "15 Aug 24",
    completedDate: "20 Feb 24",
    currency: 2800,
    status: true,
    projectName: "Test",
    fullName: "Alice Brown",
  },
];
