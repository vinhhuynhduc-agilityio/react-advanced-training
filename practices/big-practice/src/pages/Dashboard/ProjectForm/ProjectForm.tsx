import React from "react";
import { useForm } from "react-hook-form";

// types
import { ProjectsData } from "@/types/table";

interface ProjectFormProps {
  projects: ProjectsData[];
  onClose: () => void;
  onSubmit: (newProjectName: string) => void;
};

interface ProjectFormData {
  projectName: string;
};

const ProjectForm: React.FC<ProjectFormProps> = ({
  projects,
  onClose,
  onSubmit
}) => {

  const {
    register,
    handleSubmit,
    formState: { errors }
  } = useForm<ProjectFormData>();

  const isProjectDuplicate = (name: string) => {
    return projects.some(project => project.projectName === name);
  };

  const onSubmitForm = (data: ProjectFormData) => {
    onSubmit(data.projectName);
  };

  const renderError = (content: string | undefined) => (
    <p className="text-red-500 text-sm mt-1 ml-32">
      {content}
    </p>
  );

  return (
    <form onSubmit={handleSubmit(onSubmitForm)} className="space-y-4">

      {/* Body */}
      <div>
        <div className="flex items-center">
          <label className="w-32 text-gray-700 font-semibold">
            Project Name
          </label>
          <input
            className="flex-1 p-2 border border-gray-300 rounded-md bg-white text-black"
            type="text"
            {...register("projectName", {
              required: "Project name is required",
              validate: {
                duplicate: (value) => !isProjectDuplicate(value) || "Project name already exists",
              },
            })}
            placeholder="Enter project name"
          />
        </div>
        {errors.projectName && renderError(errors.projectName.message)}
      </div>

      {/* Footer */}
      <div className="flex justify-between border-t pt-3">
        <button
          type="button"
          onClick={onClose}
          className="bg-slate-200 text-pink-600 font-bold">
          Cancel
        </button>
        <button
          type="submit"
          className="bg-slate-200 text-blue-600 font-bold">
          Save
        </button>
      </div>
    </form>
  );
};

export default ProjectForm;
