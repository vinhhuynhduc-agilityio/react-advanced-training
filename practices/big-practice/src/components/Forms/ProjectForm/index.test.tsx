import { render, screen, fireEvent } from '@testing-library/react';
import { mockContextValue } from '@/mocks/data';
import ProjectForm from '.';
import { DashboardContext } from '@/context';

describe('ProjectForm', () => {
  const mockOnClose = jest.fn();
  const mockOnSubmit = jest.fn();


  const setup = () => {
    return render(
      <DashboardContext.Provider value={mockContextValue}>
        <ProjectForm
          onClose={mockOnClose}
          onSubmit={mockOnSubmit}
        />
      </DashboardContext.Provider>
    );
  };

  it('matches snapshot for default state', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });


  it('should render the project form correctly', () => {
    setup();

    expect(screen.getByText('Project Name')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Enter project name')).toBeInTheDocument();
    expect(screen.getByText('Cancel')).toBeInTheDocument();
    expect(screen.getByText('Save')).toBeInTheDocument();
  });

  it('should call onClose when the cancel button is clicked', () => {
    setup();
    fireEvent.click(screen.getByText('Cancel'));
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('should display an error when submitting with an empty project name', async () => {
    setup();
    fireEvent.click(screen.getByText('Save'));

    expect(await screen.findByText('Project name is required')).toBeInTheDocument();
    expect(mockOnSubmit).not.toHaveBeenCalled();
  });

  it('should display an error when a duplicate project name is entered', async () => {
    setup();

    const input = screen.getByPlaceholderText('Enter project name');
    fireEvent.change(input, { target: { value: 'Support' } });
    fireEvent.click(screen.getByText('Save'));

    expect(
      await screen.findByText('Project name already exists')
    ).toBeInTheDocument();
    expect(mockOnSubmit).not.toHaveBeenCalled();
  });

  it('should call onSubmit with a new project name when valid input is provided', async () => {
    setup();

    const input = screen.getByPlaceholderText('Enter project name');
    fireEvent.change(input, { target: { value: 'New Project' } });
    fireEvent.click(screen.getByText('Save'));

    await screen.findByDisplayValue('New Project');
    expect(mockOnSubmit).toHaveBeenCalledWith('New Project');
    expect(mockOnSubmit).toHaveBeenCalledTimes(1);
  });
});
