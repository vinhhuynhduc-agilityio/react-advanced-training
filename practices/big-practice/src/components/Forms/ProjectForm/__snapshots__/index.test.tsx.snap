// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectForm matches snapshot for default state 1`] = `
<div>
  <form
    class="space-y-4"
  >
    <div>
      <div
        class="flex items-center"
      >
        <label
          class="w-32 text-gray-700 font-semibold"
          for="projectName"
        >
          Project Name
        </label>
        <input
          class="flex-1 p-2 border border-gray-300 rounded-md bg-white text-black"
          id="projectName"
          name="projectName"
          placeholder="Enter project name"
          type="text"
        />
      </div>
    </div>
    <div
      class="flex justify-between border-t pt-3"
    >
      <button
        class="bg-slate-200 text-pink-600 font-bold"
        type="button"
      >
        Cancel
      </button>
      <button
        class="bg-slate-200 text-blue-600 font-bold"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
